src/message_bus.zig

```diff
@@ -18,38 +22,152 @@ pub const Message = struct {
 };
 
 pub const MessageBus = struct {
-    allocator: *Allocator,
+    const Envelope = struct {
+        message: *Message,
+        next: ?*Envelope = null,
+    };
+
+    allocator: *mem.Allocator,
     allocated: usize = 0,
-    configuration: []ConfigurationAddress,
+    io: *IO,
+    /// The replica which is running the server
+    server: *Replica,
+    server_fd: os.socket_t,
+    /// Send only connections to fellow replicas, automatically reconnected on error.
+    /// The slot for the server's replica id is undefined except for its send_queue.
+    replicas: []ReplicaConnection,
+    /// Receive only connections to fellow replicas and send/receive
+    /// connections to clients.
+    connections: []Connection,
+    /// Number of connections in the connection array currently in use
+    connections_used: usize = 0,
+    /// Number of currently active replica connections in the connections slice.
+    replica_connections: usize = 0,
 
-    /// Messages that are ready to send:
-    ready: FIFO(Envelope) = .{},
+    /// Initialize the MessageBus for the given server replica and configuration.
+    pub fn init(
+        self: *MessageBus,
+        allocator: *mem.Allocator,
+        io: *IO,
+        server: *Replica,
+        server_id: u16,
+        configuration: []std.net.Address,
+    ) !void {
+        const replicas = try allocator.alloc(ReplicaConnection, configuration.len);
+        errdefer allocator.free(replicas);
 
-    const Address = union(enum) {
-        replica: *Replica,
-    };
+        for (configuration) |address, i| {
+            if (i == server_id) {
+                replicas[i].send_queue = .{};
+            } else {
+                replicas[i] = .{ .connection = .{
+                    .message_bus = self,
+                    .address = address,
+                } };
+                replicas[i].connect();
+            }
+        }
 
-    const Envelope = struct {
-        address: Address,
-        message: *Message,
-        next: ?*Envelope,
-    };
+        const connections = try allocator.alloc(Connection, num_connections);
+        errdefer allocator.free(connections);
+        mem.set(Connection, connections, .{ .message_bus = self });
 
-    pub fn init(allocator: *Allocator, configuration: []ConfigurationAddress) !MessageBus {
-        var self = MessageBus{
+        self.* = .{
             .allocator = allocator,
-            .configuration = configuration,
+            .io = io,
+            .server = server,
+            .server_fd = try init_tcp(configuration[server_id]),
+            .replicas = replicas,
+            .connections = connections,
         };
-        return self;
+
+        self.maybe_accept();
```

Comment:
I think `maybe_accept()` should be called outside of `init()` only when the embedder is ready to start the loop. I've also got an idea on this you'll see in another comment further down for how we can decouple these kinds of calls and centralize them in one place.